packages/react-native-web/src/exports/Image/index.js

```diff
@@ -94,56 +100,105 @@ function getFlatStyle(style, blurRadius, filterId) {
   return [flatStyle, resizeMode, _filter, tintColor];
 }
 
-function resolveAssetDimensions(source) {
-  if (typeof source === 'number') {
-    const { height, width } = getAssetByID(source);
-    return { height, width };
-  } else if (
-    source != null &&
-    !Array.isArray(source) &&
-    typeof source === 'object'
-  ) {
-    const { height, width } = source;
-    return { height, width };
-  }
+function resolveAssetDimensions(source: ImageSource) {
+  const { height, width } = source;
+  return { height, width };
 }
 
-function resolveAssetUri(source): ?string {
-  let uri = null;
+function resolveSource(source: ?Source): ImageSource {
+  let resolvedSource = { uri: '' };
+
   if (typeof source === 'number') {
-    // get the URI from the packager
-    const asset = getAssetByID(source);
-    let scale = asset.scales[0];
-    if (asset.scales.length > 1) {
-      const preferredScale = PixelRatio.get();
-      // Get the scale which is closest to the preferred scale
-      scale = asset.scales.reduce((prev, curr) =>
-        Math.abs(curr - preferredScale) < Math.abs(prev - preferredScale)
-          ? curr
-          : prev
+    resolvedSource = resolveNumericSource(source);
+  } else if (typeof source === 'string') {
+    resolvedSource = resolveStringSource(source);
+  } else if (Array.isArray(source)) {
```

Comment:
I'm a little lost on this part or at least lost on why we are making the change and what problem it solves. 

Looks like we are saying here that `ImageSource` cannot be an array. And today if you pass an array as the source it looks like we will return `null`. 

Is that bad? Can we keep the same logic? Why do we want to let people pass an array at all?

Reply:
If we remove the `else if (Array.isArray(source)) {` statement
We get the following Flow errors:

```log
> npm run flow && npm run format && npm run lint && npm run unit --runInBand


> react-ui-monorepo@0.0.0 flow
> flow --flowconfig-name ./configs/.flowconfig

Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ packages/react-native-web/src/exports/Image/index.js:136:18

Property width is missing in Array [1]. [prop-missing]

 [1] 108│ function resolveSource(source: ?Source): ImageSource {
        :
     133│   } else if (typeof source === 'string') {
     134│     resolvedSource.uri = source;
     135│   } else if (source && typeof source.uri === 'string') {
     136│     const { uri, width, height, headers } = source;
     137│     resolvedSource = { uri, width, height, headers };
     138│   }
     139│


Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ packages/react-native-web/src/exports/Image/index.js:136:25

Property height is missing in Array [1]. [prop-missing]

 [1] 108│ function resolveSource(source: ?Source): ImageSource {
        :
     133│   } else if (typeof source === 'string') {
     134│     resolvedSource.uri = source;
     135│   } else if (source && typeof source.uri === 'string') {
     136│     const { uri, width, height, headers } = source;
     137│     resolvedSource = { uri, width, height, headers };
     138│   }
     139│


Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ packages/react-native-web/src/exports/Image/index.js:136:33

Property headers is missing in Array [1]. [prop-missing]

 [1] 108│ function resolveSource(source: ?Source): ImageSource {
        :
     133│   } else if (typeof source === 'string') {
     134│     resolvedSource.uri = source;
     135│   } else if (source && typeof source.uri === 'string') {
     136│     const { uri, width, height, headers } = source;
     137│     resolvedSource = { uri, width, height, headers };
     138│   }
     139│



Found 3 errors

Process finished with exit code 2
```

Previously with the `isArray` check, Flow knew that for any following `else if` it wasn't possible `Source` to be an array 

With this check removed the `else if (source && typeof source.uri === 'string') {` fails because at that point source can be an object or an array and it's not safe to destruct and use these fields 

```js
const { uri, width, height, headers } = source;
```

You can try this out by removing the statement and running `npm test` from the root project

Reply:
> Why do we want to let people pass an array at all?

Because it's possible in RN and web uses the same type definition for `Source` - https://reactnative.dev/docs/image#imagesource

> I'm a little lost on this part or at least lost on why we are making the change and what problem it solves.

1. The PR updates `ImageLoader.load` to work with `ImageSource` (in order to support headers)
2. `ImageSource` is not an array but the Image prop `source` is of type `Source` which is a wide union that can be `string`, `number`, object, array
3. To avoid guessing what shape of `source` prop is passed to the image component I've added a more specific `ImageSource` type and `resolveSource` to always resolve to something that we can work with without further guessing
4. In order to derive an object matching `ImageSource` we need to handle all possible shapes of `Source` 
5. This makes it necessary to handle the array case (instead of a fallback we can just throw an Error)